### PR TITLE
feat: Support TypeScript 5.4

### PR DESCRIPTION
**Pull request by eranhirsch**
The main thing is a fix to typing for hasAtLeast.

The fixes here are mostly a patch so that users can upgrade to typescript 5.4, proper and better fixes are planned for v2.

closes: #572

### CODE REVIEW

File: src/hasAtLeast.ts [hunk 1]
```diff
@@ -1,14 +1,41 @@
...
 
-type ArrayMinN<T, N extends number> = number extends N
-  ? // We can only compute the type for a literal number!
-    Array<T>
-  : // I don't know why we need to special-case the 0 case, but otherwise
-    // typescript complains we have a deep recursion.
-    N extends 0
-    ? Array<T>
-    : [...ReadonlyTuple<T, N>, ...Array<T>];
+type ArraySetRequired<
```
Comment (cjquines) [inline src/hasAtLeast.ts:4]: **question:** should we file an issue with type-fest?

Comment (cjquines) [inline src/hasAtLeast.ts:4]: **question:** why the rename? i think i understood `ArrayMinN` better than this

File: src/hasAtLeast.ts [hunk 2]
```diff
@@ -1,14 +1,41 @@
...
+  T extends IterableContainer,
+  N extends number,
+  Iteration extends ReadonlyArray<unknown> = [],
```
Comment (cjquines) [inline src/hasAtLeast.ts:7]: **nit:** maybe `Result`?

**Review by cjquines — APPROVED** [review]

File: src/hasAtLeast.ts (hunk 1 above)
Comment (eranhirsch) [reply src/hasAtLeast.ts:4]: We used a copy-pasted version from a few months ago; I didn't check back to see if they changed their implementation for ts5.4. Re-reading our code, I thought that building our own version made sense here.

Generally, my feelings towards type-fest shifted with the work for v2, as many type-fest types don't make sense in real-world use cases, feel more "academic"/"challenge-like"  than practical, and aren't always helpful when typing our functions. I'd rather reduce our surface area to where we use their types as-is when they fit perfectly (or we adjust the runtime behavior to match their type).

Comment (eranhirsch) [reply src/hasAtLeast.ts:4]: ArrayMinN took the **item's type** as T, just like Array does (Array<T> vs ArrayMinN<T, N>), but now the function takes the **input array** as T. It works more similarly to type-fests [SetRequired](https://github.com/sindresorhus/type-fest/blob/main/source/set-required.d.ts) where T is the object that is being transformed, and the type name describes the transformation (and not the output).

File: src/hasAtLeast.ts (hunk 2 above)
Comment (eranhirsch) [reply src/hasAtLeast.ts:7]: Oops, I was supposed to add `undefined` to this type, not the Head. Notice that we never return this type, we return T from within the recursion. This is just the regular typescript trick to stop recursions after N iterations.

**Pull request merged**